thinc/loss.py

```diff
@@ -96,26 +99,44 @@ def convert_truths(self, truths, guesses: Floats2d) -> Tuple[Floats2d, Floats2d]
                             negatives_mask[i] = 0  # type: ignore
                             negatives_mask[i][neg_index] = -1  # type: ignore
                     truths = [self._name_to_i[name] for name in truths]
+            # cast truths as 1d array
             truths = xp.asarray(truths, dtype="i")
+            # get mask for 1d array
             mask = _make_mask(guesses, missing)
         else:
+            # get mask for 2d array
             mask = _make_mask_by_value(truths, guesses, missing_value)
-        if truths.ndim != guesses.ndim:
-            # transform categorical values to one-hot encoding
-            truths = to_categorical(
-                cast(Ints1d, truths),
-                n_classes=guesses.shape[-1],
-                label_smoothing=self.label_smoothing,
+        truths = cast(IntsOrFloats, truths)
+        # Convert 1d truths to 2d and apply smoothing.
+        if truths.ndim == 1:
+            truths = cast(
+                Floats2d,
+                to_categorical(
+                    truths,
+                    n_classes=guesses.shape[-1],
+                    label_smoothing=self.label_smoothing,
+                ),
             )
-        else:
-            if self.label_smoothing:
+        # Check if 2D array is valid and maybe apply smoothing.
+        elif truths.ndim == 2:
+            # Validate if each row sums to 1.
+            if not xp.all(truths.sum(axis=1) == 1):
                 raise ValueError(
-                    "Label smoothing is only applied, when truths have type "
-                    "List[str], List[int] or Ints1d, but it seems like Floats2d "
-                    "was provided."
+                    "Cannot calculate CategoricalCrossentropy. "
+                    "All rows of 'truths' have to be a "
+                    "valid categorical distribution (sum to 1)."
                 )
+            if self.label_smoothing:
```

Comment:
Do we want to support this case? I think it is reasonable expect that if people want one-hot semantics that they pass in the gold label representation. `convert_truths` is already a large enumeration of cases. I worry that we end up supporting the cartesian product of all representations/options.

Reply:
Not sure actually. Since we do allow users to give different kinds of representations of "one correct label per sample" I think it would be more counter intuitive then not do smoothing when its not in one of the 3 options (List[str], List[int], Floats2d(onehot)). 

With the new version of the loss we actually support less kinds of things than before: we only accept rows that correspond to categorical distributions. I think we enumerated already all the options in that realm with the current code.

Reply:
What do you think?